sqe-database-access/TextRepository.cs

```diff
@@ -446,39 +951,69 @@ UserInfo                editionUser
 		/// <param name="anchorsAfter"></param>
 		/// <param name="breakNeighboringAnchors"></param>
 		/// <returns></returns>
-		public async Task<List<SignData>> CreateSignsAsync(
-				UserInfo     editionUser
-				, uint       lineId
-				, SignData   sign
-				, List<uint> anchorsBefore
-				, List<uint> anchorsAfter
-				, bool       breakNeighboringAnchors = false) => await CreateSignsAsync(
-				editionUser
-				, lineId
-				, new List<SignData> { sign }
-				, anchorsBefore
-				, anchorsAfter
-				, breakNeighboringAnchors);
+		public async Task<(List<SignData> NewSigns, List<uint>AlteredSigns)>
+				CreateSignWithSignInterpretationAsync(
+						UserInfo     editionUser
+						, uint?      lineId
+						, SignData   sign
+						, List<uint> anchorsBefore
+						, List<uint> anchorsAfter
+						, uint?      signInterpretationId
+						, bool       breakNeighboringAnchors = false
+						, bool       materializeSignStream   = true)
+			=> await CreateSignsWithInterpretationsAsync(
+					editionUser
+					, lineId
+					, new List<SignData> { sign }
+					, anchorsBefore
+					, anchorsAfter
+					, signInterpretationId
+					, breakNeighboringAnchors
+					, materializeSignStream);
 
 		/// <summary>
-		///  Join each member of firstSignInterpretations with each member of secondSignInterpretations
+		///  Join each sign interpretaton referenced by firstSignInterpretationIds
+		///  with each sign interpretation referenced by secondSignInterpretationId
 		///  in the edition sign stream.
 		/// </summary>
 		/// <param name="editionUser">The details of the edition and user</param>
-		/// <param name="firstSignInterpretations">
-		///  A list of sign interpretations to link to each member of
-		///  secondSignInterpretations
+		/// <param name="firstSignInterpretationIds">
+		///  A list of sign interpretation ids to link to each member of
+		///  secondSignInterpretationId
 		/// </param>
-		/// <param name="secondSignInterpretations">
-		///  A list of sign interpretations to be set as next sign interpretation for
-		///  each member of firstSignInterpretations
+		/// <param name="secondSignInterpretationIds">
+		///  A list of sign interpretation ids to be set as next sign interpretation for
+		///  each member of firstSignInterpretationIds
 		/// </param>
 		/// <returns></returns>
 		public async Task LinkSignInterpretationsAsync(
 				UserInfo            editionUser
-				, IEnumerable<uint> firstSignInterpretations
-				, IEnumerable<uint> secondSignInterpretations)
+				, IEnumerable<uint> firstSignInterpretationIds
+				, IEnumerable<uint> secondSignInterpretationIds
+				, bool              materializeSignStream = true)
 		{
+			// First check against cycles in the graph
+			foreach (var firstSI in firstSignInterpretationIds)
+			{
+				foreach (var secondSI in secondSignInterpretationIds)
+				{
+					var createsCycle = await _materializationRepository.IsCycleAsync(
```

Comment:
I have this half-baked idea of imposing a total order on the partial order of signs we have, which will make cycle detection extremely easy. I'm not sure it's possible in our case, Let's discuss this on Monday.